feat(progress): Add debouncing to prevent spam from rapid updates

jellingson opened the pull request:
## Add debouncing to progress plugin to prevent notification spam

### Problem
The progress plugin currently posts to Discord immediately when changes are detected. This causes spam when Brandon's website updates multiple times in quick succession (e.g., update → revert → update), flooding the Discord channel with 
notifications.

### Solution
Added a configurable debounceDelay option that waits for a specified period after the last detected change before posting. This ensures only the final state is posted after rapid changes settle.

### Changes
• **New config option**: debounceDelay (e.g., "2m", "30s") - optional, defaults to immediate posting
• **Enhanced offset format**: Stores timing and pending change information for debouncing
• **Backward compatibility**: Automatically handles legacy offset format
• **Updated documentation**: README reflects new configuration options

### Usage
yaml
connectors:
  brandon-progress:
    plugin: progress
    config:
      url: https://brandonsanderson.com
      message: The progress bars on Brandon's website were updated!
      debounceDelay: "2m"  # Wait 2 minutes after last change

### Testing
Tested with rapid successive runs (modifying the saved offset in between) - no immediate posts during changes, only posts after delay expires.

jellingson committed:
feat(progress): add debounce support with comprehensive tests

marvin-roesch commented on plugins/progress.go:
```diff
@@ -57,41 +66,78 @@ const (
+		}
+	}
+
+	// Handle legacy offset format ([]Progress) and new format (ProgressOffset)
```
This is unfortunately not gong to be sufficient for migrating the old format to the new, as at this point the JSON will already have been parsed as whatever `OffsetPrototype` returns.

I'll be thinking about this since the current approach of parsing offsets is kind of wonky anyway. Nothing to do on your part for now, I'll probably be implementing some changes that you can then rebase onto.

jellingson replied:
Changed the parsing logic to happen during `UnmarshalJSON` and validated it does work now. Reworking the offset parsing is probably a good idea. I can wait for that as well.

marvin-roesch commented on plugins/progress.go:
```diff
@@ -57,41 +66,78 @@ const (
 func (plugin ProgressPlugin) Check(offset interface{}, context PluginContext) (interface{}, error) {
 	context.Info.Println("Checking for progress updates...")
 
+	// Parse debounce delay
```
No need to perform parsing here (and if it failed, we should return the error, anyway). Instead the struct can directly use a `time.Duration` value. The parser is aware of how to parse time values.

jellingson replied:
Good to know. I'll update it in the next revision.

jellingson replied:
Updated in the latest revision.

marvin-roesch commented on plugins/progress.go:
```diff
@@ -57,41 +66,78 @@ const (
+			// Check if debounce period has passed
+			if now.Sub(*oldOffset.LastChangeTime) >= debounceDelay {
+				context.Info.Println("Debounce period passed, posting final update...")
+				if err = plugin.reportProgress(context.Discord, oldOffset.PendingDiff); err != nil {
```
There's a chance for missing entries in the diff here. We've seen Dragonsteel update one progress bar after another before. So we could have the following scenario:
 1. Bar 1 is updated -> stored as latest progress in the offset as well as a pending change
 2. Bar 2 is updated within the debounce delay -> overrides latest offset and pending change
 3. Bars stabilize -> only change in bar 2 is reported

Instead of storing the pending diff, I think it makes more sense to store the progress at the start of the debounce period and then calculate the final diff before posting.

jellingson replied:
Shoot you are right, I thought I thought I was storing that, but I messed up and overwrite the original Progress when the debounce period hasn't been met. Are you opposed if I add unit tests? I did some manual testing, but it's getting complicated haha.

jellingson replied:
Addressed in the latest revision. I also added a test case to validate this.

marvin-roesch reviewed (changes requested):
Thanks for this submission! There's only a few things I have to note, one of which will require some work on my part first (unless you also want to tackle the offset parsing).

marvin-roesch commented on plugins/progress_test.go:
```diff
@@ -0,0 +1,572 @@
+
+			// Setup Discord expectations
+			if tt.expectedPosts > 0 {
+				mockDiscord.On("Send", "Progress updated!", mock.Anything, mock.Anything, mock.Anything).Return(nil).Times(tt.expectedPosts)
```
Thanks for adding tests, I was a bit too lazy to introduce them so far, but it definitely makes sense at this level of complexity.

However, I think it would be good to actually check the content of the posted messages here to verify that the progress reports are correctly done. Could separate the actual reporting into its own struct with a method that accepts a `ProgressDiff` so that that can be mocked for easier testing, too.

jellingson replied:
Added discord message validation. I'm sure there is a better way to do it, but this is what I could come up with.

marvin-roesch commented on plugins/progress.go:
```diff
@@ -68,30 +119,67 @@ func (plugin ProgressPlugin) Check(offset interface{}, context PluginContext) (i
 	}
 
-	context.Info.Println("Reporting changed progress bars...")
+	if oldOffset.PendingChange && sameDiff(differences, oldOffset.PendingDiff) {
```
Hm, I'm not sure I quite like that the condition we have is that the *diff* be unchanged. It makes it harder to follow when exactly we're going to post a message. Hence my original suggestion to keep around some sort of `LastStableState` in the offset that we *additionally* diff against. That way, the current `Progress` value also isn't doing double-duty as both the actual last observed value and a "cache" for what it was at some point before changes started occurring.

Then the condition to enter the debounce period goes back to "the diff between the last check and the current progress is `nil`" and we reset any pending updates when the diff to the last stable state also becomes `nil`.

Finally, as a bit of a more general note: The nesting is getting a bit deep here and I generally like the "intended" effect of a function (in this case, posting the update) to be near the end. So inverting some of these conditions and turning them into early returns would be great and help readability.

I think that might also surface that `PendingChange` and `LastChangeTime` are be a bit redundant, as the former only ever is `false` when the latter is `nil` and vice-versa. Reducing the state we maintain between checks is always desirable.

jellingson replied:
> That way, the current Progress value also isn't doing double-duty as both the actual last observed value and a "cache" for what it was at some point before changes started occurring.

As a small nit, the current Progress was only intended to be the last posted value. I don't think I'm setting it to the last observed value until we decide to post it (since it would then be the last posted value). The idea is `Progress` is what you are calling `LastStableState`. I think the issue is I'm not storing the last seen change anywhere, just the last seen diff - hence the double check, Regardless, it's clearly confusing so I'll change it in the next revision.

jellingson replied:
Addressed in the next revision.

marvin-roesch reviewed (changes requested):
Thanks for these changes, especially the unit tests! I *think* the logic is all good now, but I still find a little hard to follow. See my suggestion to maybe improve that.

jellingson commented:
Made a bunch of changes to make the logic clearer. Mainly

- Changed the `ProgressOffset` struct to keep track of the `PostedState`, `ObservedState` and `DebounceStart`
    - Removed `PendingDiff` because it required both recalculating the difference between `Progress` and the website + checking if that diff was the same as the `PendingDiff`
    - Removed `PendingChange` because it is unnecessary (as you pointed out)
- Added functions to ProgressOffset to make it clear what we are checking and to remove most of the nested if statements (there is one error case when posting I couldn't figure out how to remove)
- Refactored tests to make the scenarios easier to read and understand
- Added discord message validation

jellingson commented:
Fixed a stupid mistake I made in the previous revision and added comments in the test scenarios to better explain what it is testing.
